File: backend/app/models/identity.py
import enum
from datetime import datetime
from uuid import UUID, uuid4

from sqlalchemy import BigInteger, DateTime, Enum, ForeignKey, Integer, String, func
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.dialects.postgresql import UUID as PGUUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class UserStatus(str, enum.Enum):
    new = "new"
    active = "active"
    suspended = "suspended"
    blocked = "blocked"
    deleted = "deleted"


class LifecycleEventType(str, enum.Enum):
    hire = "hire"
    transfer = "transfer"
    leave_start = "leave_start"
    leave_end = "leave_end"
    terminate = "terminate"


class LifecycleEventSource(str, enum.Enum):
    hr_system = "hr_system"
    manual = "manual"
    scheduled = "scheduled"


class LifecycleEventStatus(str, enum.Enum):
    pending = "pending"
    processed = "processed"
    failed = "failed"


class Position(Base):
    __tablename__ = "positions"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    code: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    level: Mapped[int] = mapped_column(Integer, nullable=False, default=1)


class Department(Base):
    __tablename__ = "departments"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    code: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    parent_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("departments.id"), nullable=True
    )
    manager_user_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True),
        ForeignKey("users_ext.id", use_alter=True, name="fk_department_manager"),
        nullable=True,
    )

    parent: Mapped["Department | None"] = relationship(
        "Department",
        remote_side="Department.id",
        back_populates="children",
        foreign_keys="[Department.parent_id]",
    )
    children: Mapped[list["Department"]] = relationship(
        "Department",
        back_populates="parent",
        foreign_keys="[Department.parent_id]",
    )


class UserExt(Base):
    __tablename__ = "users_ext"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    ldap_dn: Mapped[str | None] = mapped_column(String(512), unique=True, nullable=True)
    employee_id: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    username: Mapped[str] = mapped_column(String(100), unique=True, nullable=False)
    email: Mapped[str] = mapped_column(String(255), nullable=False)
    full_name: Mapped[str] = mapped_column(String(255), nullable=False)
    position_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("positions.id"), nullable=True
    )
    department_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("departments.id"), nullable=True
    )
    status: Mapped[UserStatus] = mapped_column(
        Enum(UserStatus, name="user_status"), nullable=False, default=UserStatus.active
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now(), onupdate=func.now()
    )

    position: Mapped["Position | None"] = relationship(
        "Position", foreign_keys=[position_id], lazy="selectin"
    )
    department: Mapped["Department | None"] = relationship(
        "Department", foreign_keys=[department_id], lazy="selectin"
    )
    lifecycle_events: Mapped[list["LifecycleEvent"]] = relationship(
        "LifecycleEvent", back_populates="user", foreign_keys="[LifecycleEvent.user_id]"
    )


class LifecycleEvent(Base):
    __tablename__ = "lifecycle_events"

    id: Mapped[UUID] = mapped_column(PGUUID(as_uuid=True), primary_key=True, default=uuid4)
    user_id: Mapped[UUID | None] = mapped_column(
        PGUUID(as_uuid=True), ForeignKey("users_ext.id"), nullable=True
    )
    event_type: Mapped[LifecycleEventType] = mapped_column(
        Enum(LifecycleEventType, name="lifecycle_event_type"), nullable=False
    )
    source: Mapped[LifecycleEventSource] = mapped_column(
        Enum(LifecycleEventSource, name="lifecycle_event_source"), nullable=False
    )
    payload: Mapped[dict] = mapped_column(JSONB, nullable=False, default=dict)
    processed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    status: Mapped[LifecycleEventStatus] = mapped_column(
        Enum(LifecycleEventStatus, name="lifecycle_event_status"),
        nullable=False,
        default=LifecycleEventStatus.pending,
    )
    kafka_offset: Mapped[int | None] = mapped_column(BigInteger, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )

    user: Mapped["UserExt | None"] = relationship(
        "UserExt", back_populates="lifecycle_events", foreign_keys=[user_id]
    )

File: frontend/src/api/identity.ts
import api from "./client";

export type UserStatus = "new" | "active" | "suspended" | "blocked" | "deleted";

export interface Position {
  id: string;
  code: string;
  name: string;
  level: number;
}

export interface Department {
  id: string;
  code: string;
  name: string;
  parent_id?: string;
  manager_user_id?: string;
}

export interface User {
  id: string;
  employee_id: string;
  username: string;
  email: string;
  full_name: string;
  status: UserStatus;
  position_id?: string;
  department_id?: string;
  position?: Position;
  department?: Department;
  ldap_dn?: string;
  created_at: string;
  updated_at: string;
}

export interface UsersListResponse {
  items: User[];
  total: number;
  page: number;
  page_size: number;
}

export interface LifecycleEvent {
  id: string;
  user_id?: string;
  event_type: string;
  source: string;
  status: string;
  processed_at?: string;
  created_at: string;
  payload: Record<string, unknown>;
}

export interface LifecycleEventsResponse {
  items: LifecycleEvent[];
  total: number;
}

export interface UserCreateBody {
  employee_id: string;
  username: string;
  email: string;
  full_name: string;
  position_code?: string;
  department_code?: string;
}

export interface UserUpdateBody {
  email?: string;
  full_name?: string;
  position_code?: string;
  department_code?: string;
}

export const identityApi = {
  listUsers: (params: {
    status?: string;
    search?: string;
    department_id?: string;
    page?: number;
    page_size?: number;
  }) => api.get<UsersListResponse>("/api/identity/users", { params }),

  getUser: (id: string) => api.get<User>(`/api/identity/users/${id}`),

  createUser: (body: UserCreateBody) => api.post<User>("/api/identity/users", body),

  updateUser: (id: string, body: UserUpdateBody) =>
    api.patch<User>(`/api/identity/users/${id}`, body),

  suspendUser: (id: string) => api.post<User>(`/api/identity/users/${id}/suspend`),
  restoreUser: (id: string) => api.post<User>(`/api/identity/users/${id}/restore`),
  blockUser: (id: string) => api.post<User>(`/api/identity/users/${id}/block`),
  deleteUser: (id: string) => api.delete(`/api/identity/users/${id}`),

  resetPassword: (id: string, new_password: string) =>
    api.post(`/api/identity/users/${id}/reset-password`, { new_password }),

  listPositions: () => api.get<Position[]>("/api/identity/positions"),
  listDepartments: () => api.get<Department[]>("/api/identity/departments"),

  listEvents: (params: { user_id?: string; event_type?: string; page?: number }) =>
    api.get<LifecycleEventsResponse>("/api/identity/events", { params }),
};
